crates/bls-crypto/src/bls/batch.rs

```diff
@@ -0,0 +1,86 @@
+use super::{PublicKey, Signature};
+
+// use algebra::{bls12_377::G1Projective, Field, PrimeField, ToBytes};
+use ark_bls12_377::{Fr, G1Projective};
+use ark_ff::{Field, PrimeField, ToBytes};
+
+use blake2s_simd::{Params, State};
+
+use crate::{BLSError, HashToCurve};
+
+#[derive(Default)]
+pub struct Batch {
+    entries: Vec<(PublicKey, Signature)>,
+    message: Vec<u8>,
+    extra_data: Vec<u8>,
+    size: usize,
+}
+
+impl Batch {
+    /// Constructs a new strict batch verifier context for a given message.
+    pub fn new(message: &[u8], extra_data: &[u8]) -> Batch {
+        Batch {
+            entries: Vec::<(PublicKey, Signature)>::new(),
+            message: message.to_vec(),
+            extra_data: extra_data.to_vec(),
+            size: 0,
+        }
+    }
+
+    pub fn add(&mut self, public_key: &PublicKey, signature: &Signature) {
+        self.entries
+            .push((public_key.to_owned(), signature.to_owned()));
+        self.size += 1
+    }
+
+    pub fn verify<H: HashToCurve<Output = G1Projective>>(
+        &self,
+        hash_to_g1: &H,
+    ) -> Result<(), BLSError> {
+        let mut public_keys = Vec::<&PublicKey>::new();
+        let mut signatures = Vec::<&Signature>::new();
+
+        let mut hash_states = Vec::<State>::new();
+        let mut hash_inputs = Vec::<Vec<u8>>::new();
+
+        self.entries.iter().for_each(|(pk, sig)| {
+            public_keys.push(&pk);
+            signatures.push(&sig);
+
+            hash_states.push(
+                Params::new()
+                    .personal(b"bvblssig")
+                    .hash_length(32)
```

Comment:
Why not specify `exp_size` here instead of slicing later?

Reply:
It varies based on how many you add. The idea was to potentially support streaming, which is harder to do if you have to declare a size or generate the exponents on accumulation rather than verification time.